### DISCUSSION TITLE
Working with tables in LuaJava

### DISCUSSION THREAD

**Discussion by slava110** [gudzpoz/luajava#121] [Q&A]
Hi. How to properly work with tables in LuaJava?
## Case 1:
For example, I want to get all keys or values of certain table.
In vanilla Lua you can do it using `pairs` function to iterate over table.
How to properly iterate over table on Java side though? Do I need to use run for that? Maybe there's more elegant solution?
Also maybe it's possible to get keys of table on C side properly without any iteration at all?
## Case 2:
If I want to insert into table or sort table in Lua I can use `table.sort` and `table.insert`. But there's no hooks to do that from Java side. Should I use `run` in this case too?

**Comment by osp54** [comment]
```java
for (Object entry : lua.toList(index)) {
    workWithEntry();
}
```

**Comment by gudzpoz** [comment]
A large part of this library is lightweight wrappers of [Lua C API](https://www.lua.org/manual/5.1/manual.html#3), with which you can do almost anything. So you might probably want try that first.

Anyway:

### Case 1

(You will need to push the table onto [the Lua stack](https://www.lua.org/manual/5.1/manual.html#3.1) before doing the following.)

There is mainly two ways:

1. Use [`toList`](https://gudzpoz.github.io/luajava/javadoc/party/iroiro/luajava/Lua.html#toList(int)) or [`toMap`](https://gudzpoz.github.io/luajava/javadoc/party/iroiro/luajava/Lua.html#toMap(int)).
2. Use [`next`](https://gudzpoz.github.io/luajava/javadoc/party/iroiro/luajava/Lua.html#next(int)) to iterate through tables manually.

### Case 2

Yes, you will probably want to use `run` for this. (You can also use the Lua C API, but it can get a bit tedious.)
